File: android/app/src/main/java/io/github/f4rsantos/organizer/widgets/WidgetRefresh.java
package io.github.f4rsantos.organizer.widgets;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;

import io.github.f4rsantos.organizer.R;

public final class WidgetRefresh {

    private WidgetRefresh() {}

    public static void refreshAll(Context context) {
        refresh(context, TasksWidgetProvider.class, R.id.widget_tasks_list);
        refresh(context, CalendarWidgetProvider.class, 0);
        refresh(context, KanbanWidgetProvider.class, 0);
        refresh(context, AgendaWidgetProvider.class, R.id.widget_agenda_list);
        refresh(context, GoalsWidgetProvider.class, R.id.widget_goals_list);
        refresh(context, SummaryWidgetProvider.class, 0);
        refresh(context, PomodoroWidgetProvider.class, 0);
    }

    private static void refresh(Context context, Class<?> provider, int collectionViewId) {
        Context app = context.getApplicationContext();
        AppWidgetManager manager = AppWidgetManager.getInstance(app);
        ComponentName component = new ComponentName(app, provider);
        int[] ids = manager.getAppWidgetIds(component);
        if (ids.length == 0) return;

        if (collectionViewId != 0) {
            manager.notifyAppWidgetViewDataChanged(ids, collectionViewId);
        }

        Intent intent = new Intent(app, provider);
        intent.setAction(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids);
        app.sendBroadcast(intent);
    }
}

File: android/app/src/main/java/io/github/f4rsantos/organizer/widgets/WidgetRefresh.kt
package io.github.f4rsantos.organizer.widgets

import android.appwidget.AppWidgetManager
import android.content.ComponentName
import android.content.Context
import android.content.Intent

import io.github.f4rsantos.organizer.R

object WidgetRefresh {

    fun refreshAll(context: Context) {
        refresh(context, TasksWidgetProvider::class.java, R.id.widget_tasks_list)
        refresh(context, CalendarWidgetProvider::class.java, 0)
        refresh(context, KanbanWidgetProvider::class.java, 0)
        refresh(context, AgendaWidgetProvider::class.java, R.id.widget_agenda_list)
        refresh(context, GoalsWidgetProvider::class.java, R.id.widget_goals_list)
        refresh(context, SummaryWidgetProvider::class.java, 0)
        refresh(context, PomodoroWidgetProvider::class.java, 0)
    }

    private fun refresh(context: Context, provider: Class<*>, collectionViewId: Int) {
        val app = context.applicationContext
        val manager = AppWidgetManager.getInstance(app)
        val component = ComponentName(app, provider)
        val ids = manager.getAppWidgetIds(component)
        if (ids.isEmpty()) return

        if (collectionViewId != 0) {
            manager.notifyAppWidgetViewDataChanged(ids, collectionViewId)
        }

        val intent = Intent(app, provider)
        intent.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        intent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        app.sendBroadcast(intent)
    }
}
